### DISCUSSION TITLE
Linting in source subdirectories

### DISCUSSION THREAD

**Discussion by ShatrovOA** [krvajal/vscode-fortran-support#588] [Q&A - deprecated see Fortran Discourse]
Hi all!

I cannot understand how creation of module files work with Modern Fortran.

My project structure looks like this
`mod` directory with linter output
```json
"fortran.linter.modOutput": "${workspaceFolder}/mod"
```

`src` directory contains `Precisions` module and `subdir` directory with 2 modules `TimerElemental` and `Timer`

Linter created only single module file module1.mod
`Timer` module depends on `TimerElemental` module.

`TimerElemental` detects precisions.mod and everything is OK, but `Timer` cannot detect either `TimerElemental` or `Precisions`

<img width="379" alt="image" src="https://user-images.githubusercontent.com/46355755/181050553-af3e05e4-cad7-4a52-9f6d-3ad3dc0122ca.png">

What is the correct configuration to work with subdirectories?

**Comment by gnikit** [comment]
So .mod files for the linter are produced upon file opening OR save. To get rid of that error you would have to go into `TimerElementalM` and do a save. This will generate the .mod file in `${workspaceFolder}/mod`.

I will admit that I am not a fan of this and I have been wanting to change it for a long time by "initialising" all source files found. The reasons why I haven't are:
- If the mod output directory is not set, the .mod files will be dumped into the source tree.
- If people are doing an in-source build that will override their properly generated .mod files, which we **really* don't want to do.

**Comment by ShatrovOA** [comment]
I did this several times (added empty line, saved, removed line, saved again), did not help.
I can see that every save does produce log:
```bash
["INFO" - 6:41:23 PM] Linter.arguments:
-Wall
-ffree-line-length-none
-ffixed-line-length-none
["INFO" - 6:41:23 PM] Linter.moduleOutput: -J  ${workspaceFolder}/mod
```

but .mod file is not created for files in subdirectories

**Comment by gnikit** [comment]
Have you tried also including the mod output dir to your include dirs? I suspect that the linter is unable to find the mod files.

**Comment by ShatrovOA** [comment]
Nothing changed =(

Added "${workspaceFolder}/mod"
to "fortran.linter.includePaths"
and "fortran.fortls.directories"

Edit.
Something changed

Precisions module has been detected, but TimerElemental not, since module has not been created
<img width="356" alt="image" src="https://user-images.githubusercontent.com/46355755/181114527-b1ccd2c5-2a73-4a2f-87e3-90b06539963d.png">

**Reply by gnikit** [reply]
In that case you will have to provide enough information to replicate this issue. That's not the expected behaviour or the behaviour I am experiencing.

**Reply by gnikit** [reply]
In general you need to generate all your mod files i.e. point the linter to an directory containing all the mod files from the build.

If the project does not compile then you will have mods missing, but I'm that case the linter throwing a diagnostic error is intentional.

**Reply by ShatrovOA** [reply]
Thanks, solved it. I had pointed linter and fortls to directory that contains mpi_f08 dummy functions, but ".mod" file was present in "mod" directory. I put it there and my ".mod" files started to generate.

Thanks

**Comment by ShatrovOA** [comment]
Dear @gnikit 
I have another question. This now relates to submodule procedures.
In a module a have declared procedure like this:
<img width="377" alt="image" src="https://user-images.githubusercontent.com/46355755/181644907-33cf0745-cb8f-4796-bbe7-d2a41540e3cf.png">

Now i am implementing it in a submodule:
<img width="347" alt="image" src="https://user-images.githubusercontent.com/46355755/181645029-ed25829b-fcdc-4006-b758-db16eb1523b7.png">

The problem is that neither hover nor code suggestions are available in implementation part. 
Is there any way to turn it on?

**Reply by gnikit** [reply]
I think that is related to fortls: https://gnikit.github.io/fortls and I think the issues that you are encountering are the following:

- https://github.com/gnikit/fortls/issues/108
- https://github.com/gnikit/fortls/issues/74
